plugins/auth-backend/src/providers/oauth2-proxy/provider.ts

```diff
@@ -0,0 +1,195 @@
+ */
+export type OAuth2ProxyResult<JWTPayload> = {
+  fullProfile: JWTPayload;
+  expiresInSeconds?: number;
+  accessToken: string;
+};
+
+/**
+ * @public
+ */
+export type Oauth2ProxyProviderOptions<JWTPayload> = {
+  /**
+   * Configure an auth handler to generate a profile for the user.
+   */
+  authHandler: AuthHandler<OAuth2ProxyResult<JWTPayload>>;
+
+  /**
+   * Configure sign-in for this provider, without it the provider can not be used to sign users in.
+   */
+  signIn: {
+    /**
+     * Maps an auth result to a Backstage identity for the user.
+     */
+    resolver: SignInResolver<OAuth2ProxyResult<JWTPayload>>;
+  };
+};
+
+interface Options<JWTPayload> {
+  logger: Logger;
+  signInResolver: SignInResolver<OAuth2ProxyResult<JWTPayload>>;
+  authHandler: AuthHandler<OAuth2ProxyResult<JWTPayload>>;
+  tokenIssuer: TokenIssuer;
+  catalogIdentityClient: CatalogIdentityClient;
+}
+
+/**
+ * @public
+ */
+export class Oauth2ProxyAuthProvider<JWTPayload>
+  implements AuthProviderRouteHandlers
+{
+  private readonly logger: Logger;
+  private readonly catalogIdentityClient: CatalogIdentityClient;
+  private readonly signInResolver: SignInResolver<
+    OAuth2ProxyResult<JWTPayload>
+  >;
+  private readonly authHandler: AuthHandler<OAuth2ProxyResult<JWTPayload>>;
+  private readonly tokenIssuer: TokenIssuer;
+
+  constructor(options: Options<JWTPayload>) {
+    this.catalogIdentityClient = options.catalogIdentityClient;
+    this.logger = options.logger;
+    this.tokenIssuer = options.tokenIssuer;
+    this.signInResolver = options.signInResolver;
+    this.authHandler = options.authHandler;
+  }
+
+  frameHandler(): Promise<void> {
+    return Promise.resolve(undefined);
+  }
+
+  async refresh(req: express.Request, res: express.Response): Promise<void> {
+    try {
+      const result = this.getResult(req);
+
+      const response = await this.handleResult(result);
+
+      res.status(200);
+      res.json(response);
+    } catch (e) {
+      this.logger.error(
+        `Exception occurred during ${OAUTH2_PROXY_JWT_HEADER} refresh`,
+        e,
+      );
+      res.status(401);
+      res.end();
+    }
+  }
+
+  start(): Promise<void> {
+    return Promise.resolve(undefined);
+  }
+
+  private async handleResult(
+    result: OAuth2ProxyResult<JWTPayload>,
+  ): Promise<AuthResponse<{ accessToken: string }>> {
+    const ctx = {
+      logger: this.logger,
+      tokenIssuer: this.tokenIssuer,
+      catalogIdentityClient: this.catalogIdentityClient,
+    };
+
+    const { profile } = await this.authHandler(result, ctx);
+
+    const backstageIdentity = await this.signInResolver(
+      {
+        result,
+        profile,
+      },
+      ctx,
+    );
+
+    return {
+      providerInfo: {
+        accessToken: result.accessToken,
+      },
+      backstageIdentity: prepareBackstageIdentityResponse(backstageIdentity),
+      profile,
+    };
+  }
+
+  private getResult(req: express.Request): OAuth2ProxyResult<JWTPayload> {
```

Comment:
I think it would be fine to inline this logic above, makes it a bit easier to follow the code imo

Reply:
This follows a pattern set already in other providers, i.e refresh just calls `getResult` and `handleResult` without containing much of a business logic itself. We tried here to keep the conventions set by other providers.

WDYT @dschwank

Reply:
Could flip it around and keep this but move in the logic from `extractJWT` too

Reply:
Merged `extractJWT` & `getResult` together 👍